pallets/frequency-tx-payment/src/lib.rs

```diff
@@ -527,110 +535,140 @@ impl<T: Config> core::fmt::Debug for ChargeFrqTransactionPayment<T> {
+		Ok((validity, val, origin))
 	}
 
-	/// Do any pre-flight stuff for a signed transaction.
-	#[allow(deprecated)]
-	fn pre_dispatch(
+	fn prepare(
 		self,
-		who: &Self::AccountId,
-		call: &Self::Call,
-		info: &DispatchInfoOf<Self::Call>,
+		val: Self::Val,
+		_origin: &<T as frame_system::Config>::RuntimeOrigin,
+		call: &<T as frame_system::Config>::RuntimeCall,
+		info: &DispatchInfoOf<<T as frame_system::Config>::RuntimeCall>,
 		len: usize,
 	) -> Result<Self::Pre, TransactionValidityError> {
-		let (_fee, initial_payment) = self.withdraw_fee(who, call, info, len)?;
-
-		Ok((self.tip(call), who.clone(), initial_payment))
+		match val {
+			Val::Charge { tip, who, .. } => {
+				let (_fee, initial_payment) = self.withdraw_fee(&who, call, info, len)?;
+				Ok(Pre::Charge { tip, who, initial_payment, weight: self.weight(call) })
+			},
+			Val::NoCharge => Ok(Pre::NoCharge { refund: self.weight(call) }),
+		}
 	}
 
-	/// Do any post-flight stuff for an extrinsic.
-	#[allow(deprecated)]
-	fn post_dispatch(
-		maybe_pre: Option<Self::Pre>,
-		info: &DispatchInfoOf<Self::Call>,
-		post_info: &PostDispatchInfoOf<Self::Call>,
+	fn post_dispatch_details(
+		pre: Self::Pre,
+		info: &DispatchInfoOf<<T as frame_system::Config>::RuntimeCall>,
+		post_info: &PostDispatchInfoOf<<T as frame_system::Config>::RuntimeCall>,
 		len: usize,
 		result: &DispatchResult,
-	) -> Result<(), TransactionValidityError> {
-		if let Some((tip, who, initial_payment)) = maybe_pre {
-			match initial_payment {
-				// If this is a Token transaction, passthrough
+	) -> Result<Weight, TransactionValidityError> {
+		match pre {
+			Pre::Charge { tip, who, initial_payment, weight } => match initial_payment {
 				InitialPayment::Token(already_withdrawn) => {
-					// post_dispatch_details eliminated the Option from the first param.
-					// TransactionExtension implementers are expected to customize Pre to separate signed from unsigned.
-					// https://github.com/paritytech/polkadot-sdk/pull/3685/files?#diff-be5f002cca427d36cd5322cc1af56544cce785482d69721b976aebf5821a78e3L875
+					let actual_ext_weight =
+						<T as Config>::WeightInfo::charge_tx_payment_token_based();
+					let unspent_weight = weight.saturating_sub(actual_ext_weight);
+					let mut actual_post_info = *post_info;
+					actual_post_info.refund(unspent_weight);
 					pallet_transaction_payment::ChargeTransactionPayment::<T>::post_dispatch_details(
                         pallet_transaction_payment::Pre::Charge { tip, who, imbalance: already_withdrawn },
                         info,
                         post_info,
                         len,
                         result,
                     )?;
+					Ok(unspent_weight)
 				},
-				// If it's capacity, do nothing
 				InitialPayment::Capacity => {
 					debug_assert!(tip.is_zero(), "tip should be zero for Capacity tx.");
+					// TODO:: Currently no refund for Capacity transactions.
+					Ok(Weight::zero())
```

Comment:
Consuming all the weight for extension for capacity payment, also capacity mechanism does not have a refund mechanism (probably not needed for capacity)